File: packages/typescript/src/classifiers/llmClassifier.ts
import type { ClassificationResult, Classifier } from "./base.ts";
import { DEFAULT_MODEL } from "../defaults.ts";
import type { LLMClient } from "../llm/client.ts";
import { LiteLLMClient } from "../llm/client.ts";
import { buildClassifierResponseSchema } from "../personas/schema.ts";
import { matchLabel, parseConfidence, safeJsonObject, stripMarkdown, wrapUntrusted } from "../utils.ts";

export type LLMClassifierOptions = {
  model?: string;
  labels?: string[];
  systemPrompt?: string;
  temperature?: number;
  llmClient?: LLMClient;
};

export class LLMClassifier implements Classifier {
  public labels: string[];
  private model: string;
  private systemPrompt: string;
  private temperature: number;
  private llmClient: LLMClient;

  constructor(options: LLMClassifierOptions = {}) {
    const cleaned = (options.labels ?? [])
      .map((label) => String(label).trim())
      .filter((label) => label.length > 0);
    if (cleaned.length === 0) {
      throw new Error(
        "LLMClassifier requires at least one non-empty label. " +
          "Pass labels: ['safe', 'unsafe'] (or similar) when constructing.",
      );
    }
    this.model = options.model ?? DEFAULT_MODEL;
    this.labels = cleaned;
    this.systemPrompt =
      options.systemPrompt ?? "Classify the text and return JSON with fields label and confidence.";
    this.temperature = options.temperature ?? 0;
    this.llmClient = options.llmClient ?? new LiteLLMClient();
  }

  /**
   * Classify `text` with one LLM call. Output the jury cannot trust (not
   * JSON, a label outside `labels`, or a non-numeric confidence) comes back
   * with confidence 0 so the jury escalates it; `rawOutput.error` says why.
   */
  async classify(text: string): Promise<ClassificationResult> {
    const prompt = [
      "Classify the following text.",
      `Labels: ${this.labels.join(", ") || "any"}`,
      `Text:\n${wrapUntrusted(text)}`,
      "Respond ONLY with JSON: {\"label\":\"...\",\"confidence\":0.0-1.0}",
    ].join("\n");

    const responseFormat = buildClassifierResponseSchema(this.labels) as unknown as Record<string, unknown>;
    const payload = await this.llmClient.complete(
      this.model,
      this.systemPrompt,
      prompt,
      this.temperature,
      responseFormat,
    );
    const parsed = safeJsonObject(stripMarkdown(payload.content));
    const fallbackLabel = this.labels[0] ?? "unknown";
    const callCost = payload.costUsd ?? null;
    if (!parsed) {
      return {
        label: fallbackLabel,
        confidence: 0,
        rawOutput: { raw_content: payload.content, error: "invalid_json" },
        costUsd: callCost,
      };
    }

    const label = matchLabel(parsed.label, this.labels);
    if (label === null) {
      return {
        label: fallbackLabel,
        confidence: 0,
        rawOutput: { raw_content: payload.content, error: "label_not_in_labels" },
        costUsd: callCost,
      };
    }

    const confidence = parseConfidence(parsed.confidence);
    if (confidence === null) {
      return {
        label,
        confidence: 0,
        rawOutput: { raw_content: payload.content, error: "invalid_confidence" },
        costUsd: callCost,
      };
    }

    return {
      label,
      confidence,
      rawOutput: parsed,
      costUsd: callCost,
    };
  }

  async classifyBatch(texts: string[]): Promise<ClassificationResult[]> {
    const out: ClassificationResult[] = [];
    for (const text of texts) {
      out.push(await this.classify(text));
    }
    return out;
  }
}

File: packages/python/src/llm_jury/classifiers/llm_classifier.py
from __future__ import annotations

from llm_jury._defaults import DEFAULT_MODEL
from llm_jury.llm.client import LiteLLMClient, LLMClient
from llm_jury.personas.schema import build_classifier_response_schema
from llm_jury.utils import (
    match_label,
    parse_confidence,
    payload_cost,
    safe_json_parse,
    strip_markdown_fences,
    wrap_untrusted,
)

from .base import ClassificationResult, Classifier


class LLMClassifier(Classifier):
    def __init__(
        self,
        model: str = DEFAULT_MODEL,
        labels: list[str] | None = None,
        system_prompt: str | None = None,
        llm_client: LLMClient | None = None,
        temperature: float = 0.0,
    ) -> None:
        cleaned = [str(label).strip() for label in (labels or []) if str(label).strip()]
        if not cleaned:
            raise ValueError(
                "LLMClassifier requires at least one non-empty label. "
                "Pass labels=['safe', 'unsafe'] (or similar) when constructing."
            )
        self.model = model
        self.labels = cleaned
        self.system_prompt = (
            system_prompt
            or "Classify the text and return JSON with label and confidence."
        )
        self.llm_client = llm_client or LiteLLMClient()
        self.temperature = temperature

    async def classify(self, text: str) -> ClassificationResult:
        """Classify ``text`` with one LLM call.

        Unusable output never raises: invalid JSON, a label outside
        ``labels`` or an invalid confidence returns confidence 0.0 (so the
        Jury escalates) with the reason in ``raw_output["error"]``.
        """
        prompt = (
            "Classify the following text into one of the available labels.\n"
            f"Labels: {', '.join(self.labels) if self.labels else 'any'}\n"
            f"Text:\n{wrap_untrusted(text)}\n"
            'Respond with JSON: {"label":"...","confidence":0.0-1.0}.'
        )
        payload = await self.llm_client.complete(
            model=self.model,
            system_prompt=self.system_prompt,
            prompt=prompt,
            temperature=self.temperature,
            response_format=build_classifier_response_schema(self.labels),
        )
        raw_content = str(payload.get("content") or "")
        call_cost = payload_cost(payload)
        data = safe_json_parse(strip_markdown_fences(raw_content))
        fallback_label = self.labels[0] if self.labels else "unknown"
        if data is None:
            return ClassificationResult(
                label=fallback_label,
                confidence=0.0,
                raw_output={"raw_content": raw_content, "error": "invalid_json"},
                cost_usd=call_cost,
            )

        label = match_label(data.get("label"), self.labels)
        if label is None:
            return ClassificationResult(
                label=fallback_label,
                confidence=0.0,
                raw_output={"raw_content": raw_content, "error": "label_not_in_labels"},
                cost_usd=call_cost,
            )

        confidence = parse_confidence(data.get("confidence"))
        if confidence is None:
            return ClassificationResult(
                label=label,
                confidence=0.0,
                raw_output={"raw_content": raw_content, "error": "invalid_confidence"},
                cost_usd=call_cost,
            )

        return ClassificationResult(
            label=label,
            confidence=confidence,
            raw_output=data,
            cost_usd=call_cost,
        )
